tests/test_instruction_tree.gd

```diff
@@ -63,5 +79,140 @@ func test_tree_node_text():
 	root.next = sibling
 	sibling.add_child(nephew)
 
-	var text: String = tree.generate_text(root, 0)
+	var text: String = InstructionTree.generate_text(root, 0)
 	assert_eq(text, "root\n\tchild1\n\t\tgrandchild\n\tchild2\nsibling\n\tnephew\n")
+
+
+func test_script_no_nodes():
+	var script = InstructionTree.generate_script_from_nodes([], "Foo")
+	assert_eq(
+		script,
+		(
+			"""\
+			extends Foo
+
+			var VAR_DICT := {}
+
+			"""
+			. dedent()
+		)
+	)
+
+
+func test_script_no_entry_blocks():
+	var nodes: Array[Node] = [Node.new(), Node2D.new(), Control.new()]
+	var script = InstructionTree.generate_script_from_nodes(nodes, "Foo")
+	assert_eq(
+		script,
+		(
+			"""\
+			extends Foo
+
+			var VAR_DICT := {}
+
+			"""
+			. dedent()
+		)
+	)
+	for node in nodes:
+		node.free()
+
+
+func test_basic_script():
+	var ready_block: Block = dup_node(general_blocks["ready_block"])
+
+	var print_block: Block = dup_node(general_blocks["print"])
+	# XXX: It seems like this should substitute {text} in the statement,
+	# but it doesn't. I can't make sense of StatementBlock.
+	# print_block.param_input_strings = {"text": "this is a test"}
+	# print_block._ready()
```

Comment:
I puzzled over this for a little bit, and this error seems pertinent:

```
SCRIPT ERROR: Invalid set index 'text' (on base: 'Nil') with value of type 'String'.
          at: ParameterInput.set_raw_input (res://addons/block_code/ui/blocks/utilities/parameter_input/parameter_input.gd:57)
```

`GutTest` is a `Node`, so we can call `add_child(ready_block)` instead of having to call `print_block._ready()` and I think that moves things a little closer. I also changed this…

```
	ready_block.bottom_snap.insert_snapped_block(print_block)
	ready_block.bottom_snap.snapped_block = print_block
```

to…

```
	ready_block.bottom_snap.add_child(print_block)
```

… but I still ran into an issue where `SnapPoint._on_child_entered_tree` isn't being called, even though SnapPoint is definitely initialized.

Reply:
Yeah, I completely missed the errors when I was running it until today :roll_eyes: I have a refactor in `ParameterInput` that allows it to work without the UI components but it's a bit messy. Since I have to rebase this after #107, I'll try to clean it up and add it.

Reply:
I did get this working in tests, but I was unsure whether I was breaking the option handling, so I'm going to punt on that part for the moment.

Reply:
```gdscript
print_block.param_input_strings = {"text": "this is a test"}
add_child(print_block) # Add print_block to scene tree to have it initialize properly
remove_child(print_block) # Make sure it's available to snap to the ready_block

ready_block.bottom_snap.insert_snapped_block(print_block) # Attach to ready_block
ready_block.bottom_snap._update_snapped_block_from_children() # Set snapped_block in ready_block
assert_true(ready_block.bottom_snap.has_snapped_block())
assert_eq(ready_block.bottom_snap.get_snapped_block(), print_block)
```

Changing the test to this works, but it's pretty weird, see the explanation of why it works in the comments. What Dylan said is exactly why the error is happening, but this is a solution. Once the new serialization is in, this test will look more clean.

Reply:
Oh, duh. I had convinced myself that `_ready` would never be called in these tests even though I knew I'd seen it run before in these contexts. Turns out GUT [has a helper](https://gut.readthedocs.io/en/latest/Memory-Management.html#using-add-child-in-tests) to add the child and have it automatically removed and freed at the end of the test.

On the other hand, I was make `ParameterInput` work outside of a scene by storing `raw_input` in the object and reworking when it's transferred to and from the UI controls. I'll try to polish that up, but if I can't I'll do it this way.